138676271 post account memberships v5

wlffann opened the pull request:
In implementing the `POST /accounts/:account_id/memberships` endpoint, this PR creates a Proxy class to handle the chained methods on `AccountMembership`. 

Merge conflicts need to be resolved.

[#138676271]

tmikeschu commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -0,0 +1,27 @@
+class PivotalTracker::AccountMembership
+  attr_accessor :kind, :id, :person, :account_id, 
```
@wlffann I think these should be alphabetized, to follow the existing templates.

jasonnoble replied:
Agreed.

wlffann replied:
Got it! Thanks

tmikeschu commented on lib/pivotal-tracker/account.rb:
```diff
@@ -0,0 +1,23 @@
+class PivotalTracker::Account
+  attr_accessor :created_at, :days_left, :id, :kind,
```
@wlffann the order for these will get updated if we merge #5 first.

jasonnoble replied:
👍

tmikeschu commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -0,0 +1,27 @@
+        PivotalTracker::AccountMembership.new(membership)
+      end
+    else
+      puts response.inspect
```
@wlffann if we merge #5 first, this error should grab the constant that will be in the PivotalTracker model.

tmikeschu commented on lib/pivotal-tracker/account_membership.rb.orig:
```diff
@@ -0,0 +1,31 @@
+
+  def self.create(membership_attributes, owner_id)
+    response = PivotalTracker.post("/accounts/#{owner_id}/memberships", query: membership_attributes)
+<<<<<<< HEAD
```
@wlffann merge conflicts! I see this has a `.orig` extension. Maybe just delete the files with that extension?

jasonnoble replied:
Yup.  Need to remove them.

wlffann replied:
Right! Thank you!

tmikeschu commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,29 @@
+class BasicObject #:nodoc:
```
@wlffann Killer trailblazing on this!

lucyconklin replied:
@wlffann @tmikeschu Yes! Thanks for tackling this. It seems like a pretty slick way to handle objects.

jasonnoble replied:
Do we need the `BasicObject` class?  See built in one:  https://ruby-doc.org/core-2.2.0/BasicObject.html

jasonnoble replied:
https://ruby-doc.org/stdlib-2.2.1/libdoc/delegate/rdoc/SimpleDelegator.html

tmikeschu commented on spec/lib/pivotal-tracker/account_spec.rb:
```diff
@@ -0,0 +1,57 @@
+  describe "#find" do
+    let(:account) { subject.find(account_id) }
+    context "with a valid account id" do
+      let(:account_id) { 962655 }
```
@wlffann I had to replace the hard-coded id's with `ENV['PIVOTAL_ACCOUNT_ID']`

jasonnoble replied:
Rebasing might fix?

tmikeschu commented on lib/pivotal-tracker/proxy.rb.orig:
```diff
@@ -0,0 +1,41 @@
+class BasicObject #:nodoc:
```
Same as comment above about deleting these `.orig` files.

tmikeschu reviewed (commented):
Stellar work! 👌

kheppenstall commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,29 @@
+    def method_missing(method, *args, &block)
+      @target.send(method, *args, &block)
+    end
+end
```
@wlffann With the `attr_reader` do we need all the `@` symbols in this file?

jasonnoble replied:
👍

jasonnoble commented on lib/pivotal-tracker/account_membership.rb.orig:
```diff
@@ -0,0 +1,31 @@
+class PivotalTracker::AccountMembership
```
`.orig` files are a side affect of `git mergetool` and should not be committed.  This is usually caused by running `git add .` instead of specifically staging the changes you want staged.

Please `git rm` this file, commit, then squash it into the existing commits.

jasonnoble commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def all
+    @target.all(@owner.id)
```
`target`

jasonnoble commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def find(param)
+    @target.find(param, @owner.id)
```
`target`

jasonnoble commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def create(args = {})
+    @target.create(args, owner.id)
```
`target`

jasonnoble commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,29 @@
+
+  private
+    def method_missing(method, *args, &block)
+      @target.send(method, *args, &block)
```
`target`

jasonnoble commented on lib/pivotal_tracker.rb:
```diff
@@ -7,6 +7,9 @@ class PivotalTracker
 require 'pivotal-tracker/client'
 require 'pivotal-tracker/story'
 require 'pivotal-tracker/project'
+require 'pivotal-tracker/account'
```
Please switch to `autoload`

jasonnoble commented on spec/lib/pivotal-tracker/account_spec.rb:
```diff
@@ -0,0 +1,57 @@
+    end
+
+    it "a new membership can be created" do
+      new_membership = account.memberships.create(:email => "lucy.conklin@gmail.com", :name => "Lucy Conklin", :initials => "LC")
```
Use `Faker` gem to generate a name/email/initials?  Then verify the new_membership fields?

jasonnoble commented on spec/lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,11 @@
+class Proxy
```
https://ruby-doc.org/stdlib-2.2.1/libdoc/delegate/rdoc/SimpleDelegator.html ?

jasonnoble commented on README.md.orig:
```diff
@@ -0,0 +1,73 @@
+## pivotal-tracker.rb - the original ruby gem
```
Please remove me...

jasonnoble commented on lib/pivotal-tracker/account.rb.orig:
```diff
@@ -0,0 +1,38 @@
+class PivotalTracker::Account
```
Remove me...

wlffann committed:
Implement `POST /account/:id/membership`

These changes add the `#create` method on AccountMemberships that is called through an instance of `@account`. It implements a proxy model to handle chained method calls (`@account.memberships.create()`).

[#138676271]

jasonnoble commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -1,6 +1,29 @@
-class PivotalTracker::AccountMembership < PivotalTracker::Resource
+class PivotalTracker::AccountMembership
```
Please undo this change.

jasonnoble commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -1,6 +1,29 @@
+  end
+
+  def self.create(membership_attributes, account_id)
+    response = PivotalTracker.post("/accounts/#{account_id}/memberships", query: membership_attributes)
```
Should we check for a 200 here?  (Like line 16)

jasonnoble commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,17 @@
+class PivotalTracker::Proxy
```
Inherit from resource?

wlffann replied:
Its initialize process does not take a hash of attributes, so I don't believe that it should inherit from resource

jasonnoble commented on lib/pivotal-tracker/proxy.rb:
```diff
@@ -0,0 +1,17 @@
+class PivotalTracker::Proxy
```
Is this a general purpose proxy?  Or is it specific to Account Memberships?

wlffann replied:
It is a general purpose Proxy. Any nested accounts can use to to called chained methods

jasonnoble commented on lib/pivotal_tracker.rb:
```diff
@@ -1,9 +1,9 @@
-
-class PivotalTracker
-
+require 'httparty'		
```
Unrelated change?

jasonnoble commented on lib/pivotal_tracker.rb:
```diff
@@ -13,11 +13,13 @@ class PivotalTracker
 
-class PivotalTracker
-  include HTTParty
+class PivotalTracker		
```
Remove trailing white space...

jasonnoble commented on spec/lib/pivotal-tracker/account_spec.rb:
```diff
@@ -1,52 +1,59 @@
     end
-  end
-end
+  end		
```
Trailing white space

jasonnoble commented on README.md:
```diff
@@ -66,4 +66,4 @@ The API is based on the following gist: http://gist.github.com/283120
 * Jon Mischo (https://github.com/supertaz)
-* Gabor Ratky (https://github.com/rgabo)
\ No newline at end of file
+* Gabor Ratky (https://github.com/rgabo)
```
Unrelated change...

jasonnoble commented on lib/pivotal-tracker/account.rb:
```diff
@@ -1,3 +1,4 @@
+
```
Remove me

jasonnoble commented on lib/pivotal_tracker.rb:
```diff
@@ -6,6 +6,7 @@ class PivotalTracker
   base_uri 'https://www.pivotaltracker.com/services/v5'
 
   PermissionDenied = Class.new(Error)
+
```
Unrelated...

jasonnoble commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -3,4 +3,24 @@ class PivotalTracker::AccountMembership < PivotalTracker::Resource
+    if response.code == 200
+      new_account = PivotalTracker::AccountMembership.new(response.parsed_response)
+    else
+      raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can add account memeberships")
```
`memeberships` vs `memberships`
